src/H5E.c

```diff
@@ -245,6 +245,20 @@ H5E_term_package(void)
         nstk = H5I_nmembers(H5I_ERROR_STACK);
 
         if ((ncls + nmsg + nstk) > 0) {
+            /* Clear the default error stack. Note that
+             * the following H5I_clear_type calls do not
+             * force the clears and will not be able to
+             * clear any error message IDs that are still
+             * in use by the default error stack unless we
+             * clear that stack manually.
+             *
+             * Error message IDs will typically still be
+             * in use by the default error stack when the
+             * application does H5E_BEGIN/END_TRY cleanup
+             * at the very end.
+             */
+            H5E_clear_stack(NULL);
```

Comment:
When an application does the usual H5E_BEGIN/END_TRY cleanup at the very end, the _current_ behavior in HDF5 actually relies (probably unknowingly) on the fact that H5Eauto_is_v2 clears the error stack. When you do something like:

`H5E_BEGIN_TRY {
    H5Fclose(fid);
} H5E_END_TRY;`

at the end of an application, the error message from that failure when fid is invalid gets placed on the error stack. Then, when the library is closing down, the first package that gets terminated (currently H5F in my testing) makes a call to H5I_dec_type_ref -> H5I__destroy_type which contains the following:

`H5E_BEGIN_TRY
{
    H5I_clear_type(type, TRUE, FALSE);
}
H5E_END_TRY /* don't care about errors */`

This block has the side effect of clearing the error stack, which is why the error message from the application's cleanup code doesn't get printed on library shutdown.

Modifying H5Eauto_is_v2 to NOT clear the error stack means these error messages get left inside the default error stack and prevent the H5E shutdown code from closing the error stack since the below H5I_clear_type calls don't force the type clearing. Therefore, when the H5E package is terminating and we have error message/class IDs laying around, we now have to manually clear the default error stack. However, I think this is better than relying on one of the packages to accidentally clear the stack.